### ISSUE TITLE
iframe support of H5P-Nodejs-library application in angular 6

### ISSUE THREAD

**Issue by badrinathpathak** [Lumieducation/H5P-Nodejs-library#268]
**Describe the bug**

I am trying to integrate the H5P editor in the iframe. 
H5P-Nodejs-library application is running fine and I am able to perform create and play operations,   but the issue occurs when I am trying to embed inside an iframe, it's unable to load required libraries when it embedded inside an iframe.    
UI application is written in Angular 6 and H5P-Nodejs-library is deployed in ubuntu:18.04.  

**Sever Environment (please complete the following information):**
 - OS: Ubuntu:18.04
 - Nodejs-Version v10.17.0

**Client Environment (please complete the following information):**
 - OS: Windows 10
 - Browser chrome: v78
 - Nodejs-Version v8.11.2

**To Reproduce**
1. Download H5P-Nodejs-library repo in Ubuntu, install npm dependencies, build the source code and start the server on the default port: 8080.
2. Goto client machine, where you have written angular/html code, include the H5P-Nodejs-library application URL in the iframe.
3. Goto browser and open UI URL.
4. Open the console/inspect then you can find "cross-origin" error and script loading failure messages.

**Expected behavior**
Expecting a similar web page should be loaded as in H5P-Nodejs-library application URL in iframe page also.

**Screenshots**
Image showing integration in iframe.
![1574072355415blob](https://user-images.githubusercontent.com/11488132/69045649-47041280-0a1d-11ea-9e0c-1146fdb2c784.jpg)

Image showing error in loading scripts and cross-origin.
![Capture](https://user-images.githubusercontent.com/11488132/69046206-710a0480-0a1e-11ea-8add-974beb6a715f.PNG)

**Comment by sr258** [comment]
I've looked into this issue and can't offer an out-of-the-box solution yet, but I've got it working in a crude patch in [this](https://github.com/Lumieducation/H5P-Nodejs-library/tree/try/iframe-editor) branch

To fix the CORS error, you'll have to serve both the editor and your Angular app from the same domain. Otherwise, browsers reject the AJAX calls made within the editor. A quick snippet you can put into express.ts:

```JS

server.get('/iframe', (req, res) => {
        res.send(`
        <html>
            <head></head>
            <body>
                <iframe src="/edit"></iframe>
            </body>
        </html>
        `);
    });
```

If you open the http://XXX/iframe URL you'll no longer see the CORS errors.

It still won't work then, as the H5PIntegration object is not correctly filled with the needed data. Your surrounding window (the one returned by /iframe) needs to have the H5PIntegration object with all necessary values. This is the caused by [line 11 in Joubel's h5peditor.js](https://github.com/h5p/h5p-editor-php-library/blob/91230988a00a09941ecf6225aed9549c9fa34070/scripts/h5peditor.js#L11), as the integration object of the editor is always set to the integration object of the window's **parent**.

You'll also need to include several .js scripts in the parent window and call some initialization function.

In a nutshell, this means your parent window must contain most of what you see in src/renderers/default.ts

Check out the branch I've mentioned above for more reference. I'll soon try to add documentation on how to work with iframes and integrate the necessary changes in the codebase in a better way. At this point, you can't work with iFrames without us patching the library (to make H5PEditor.integration public).

**Comment by sr258** [comment]
After a good night's sleep, I don't think it makes sense to integrate the editor into Angular like that. The h5p editor client expects all the stuff I've mentioned above to be present in the parent window, so what you basically do is to rewrite all the functionality of the renderer just to have this. It also makes sense to have the H5PIntegration in the parent window, as the iframe must know which contentId it is supposed to edit and you must have a way of communication the contents of the editor window back to the parent to save the piece of content. Furthermore, the h5p editor client already creates an iframe of its own anyway, so it's not really worth the trouble to have your own iframe.

So I'd suggest this:

- Write your own custom renderer that is based on src/renderers/default.ts and pass it to H5PEditor.useRenderer(...) after H5PEditor has been constructed. (I'll add documentation on how to write renderers soon.) You basically just have to strip the <html> and <head> tags and move the <script> tags to the body.
- Implement an endpoint in your application that calls the renderer like this: ```h5pEditor.render(contentId)``` and return its return value. (The details depend on how you serve files and requests in your application. Check out [examples/express.ts:191-193](https://github.com/Lumieducation/H5P-Nodejs-library/blob/3e24178345d74489bb8a66d47abe37677ee8698b/examples/express.ts#L191) as reference.
- In Angular, you call the endpoint from above with an AJAX call, parse the result you get and add this to the DOM in the place you want the editor to appear (where you've put the iframe).

This should get the editor working. @JPSchellenberg do you agree?

**Comment by krishna-pwc** [comment]
As per your suggestions I have written HTTP GET Call in Angular 6 to get the response from **/edit** path and I put that data into html div by using **[innerHtml]** property. But it's not working as we expected. It's showing only the create button and it's not showing any other h5p related pages.

**Environment Details**
              I have deployed the **H5P-Nodejs-library** into separate VM(Virtual Mahine) and I resolved the cors origin issue by adding the proxy. Now the only issue is like I was not able to render the H5P-Nodejs-library in Angular 6 by using iFrame or Ajax calls.

**Sever Environment**

- OS: Ubuntu:18.04
- Nodejs-Version v10.17.0

**Client Environment**

- OS: Windows 10
- Browser chrome: v78
- Nodejs-Version v8.11.2

**To Reproduce**                

1. Download H5P-Nodejs-library repo in Ubuntu, install npm dependencies, build the source code and start the server on the default port: 8080.
2. Goto client machine, where you have written **AJAX/HTTP API** Calls, include the response into the div.
3. Goto browser and open UI URL.
4. Now you can see the h5p plugin was loaded partially into the webpage.


**Screenshots**
![3](https://user-images.githubusercontent.com/40989718/69523083-fe9ea480-0f88-11ea-97a6-2ef0e0751fb1.png)
![2](https://user-images.githubusercontent.com/40989718/69523084-ff373b00-0f88-11ea-8c7c-1df2e362f896.png)
![1](https://user-images.githubusercontent.com/40989718/69523085-ff373b00-0f88-11ea-8ea8-8f5ad12b8252.png)

**Comment by sr258** [comment]
Does the browser console show any JavaScript errors? These might be helpful as a starting point for debugging. How is the response in the endpoint /h5p/content/create generated? Do you use a custom renderer or the default one?

**Comment by krishna-pwc** [comment]
Basically in browser console it's not showing any error. And this **/h5p/content/create** path has been created by using proxy. Please refer the screenshot for further information.
![4](https://user-images.githubusercontent.com/40989718/69525386-03b22280-0f8e-11ea-83d3-c7d69239f125.png)
![Capture](https://user-images.githubusercontent.com/40989718/69525388-03b22280-0f8e-11ea-8c90-d449fe8dfac5.PNG)

**Comment by sr258** [comment]
I'm not well-versed in Angular, so I'm guessing here:

It looks like the scripts required to load the h5p editor aren't loaded and executed. It's possible that inserting <script> tags into innerHTML doesn't work in Angular or in general. Maybe [this page](http://blog.davidjs.com/2018/09/insert-script-tags-in-angular-components/) is of help in that respect.

The default renderer returns a full web page with <html> and <head> tags. It's possible that his won't work with innerHTML. That's why you might have to writer your own renderer (see post above):

- Write your own custom renderer that is based on src/renderers/default.ts and pass it to H5PEditor.useRenderer(...) after H5PEditor has been constructed. You basically just have to strip the and tags and move the <script> tags to the body.
- Implement an endpoint in your application that calls the renderer like this: h5pEditor.render(contentId) and return its return value. (The details depend on how you serve files and requests in your application. Check out examples/express.ts:191-193 as reference.

**Comment by krishna-pwc** [comment]
Ok. I'll try to write my custom renderer method as you mentioned.

**Comment by JPSchellenberg** [comment]
From my point of view, the implementation via iframe is not how it is intended from joubel/h5p. 
I think the best way is to include all editor scripts and then initialize the editor in a div. That's what [the renderer](https://github.com/Lumieducation/H5P-Nodejs-library/blob/master/src/renderers/default.ts) is doing, but you can do it on every page. The renderer script is from the H5P WordPress plugin.

If all editor-scripts are loaded you just need to call 
` h5peditor = new ns.Editor(undefined, undefined, $editor[0]);` 
(See [here](https://github.com/Lumieducation/H5P-Nodejs-library/blob/master/src/renderers/default.ts#L71), where the $editor[0] is the div selected by jQuery and the editor is initialized and rendered in the div. 

The `H5PEditor.render()` method is actually mostly in here because we use it in our example. We will update the docs on how to integrate the editor better.

**Comment by sr258** [comment]
@krishna-pwc Did you succeed with the custom renderer?

**Comment by krishna-pwc** [comment]
Yes I have succeeded with the custom renderer. Thanks for your replies. 

> @krishna-pwc Did you succeed with the custom renderer?

**Comment by sr258** [comment]
Great to here that!
